File: py/695. Max Area of Island.py
def dfs(grid,i,j):
    count=0
    if(i<0 or j<0 or i>=len(grid) or j>= len(grid[0]) or grid[i][j]==0):
        return 0
    else:
        grid[i][j]=0
        count=1
        count+=dfs(grid,i+1,j)
        count+=dfs(grid,i-1,j)
        count+=dfs(grid,i,j+1)
        count+=dfs(grid,i,j-1)
    
    
    return count
    
class Solution(object):
    def maxAreaOfIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        maxi=0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if(grid[i][j]==1):
                    maxi=max(dfs(grid,i,j),maxi)
        return maxi

File: cpp/695_Max_Area_of_Island.cpp
class Solution {
public:
    int m,n;
    vector<vector<int>> a,vis;
    int rec(int i, int j)
    {
        if(i<0 || j<0 || i>=m || j>=n || a[i][j]==0 ||vis[i][j]==1)
            return 0;
        vis[i][j] = 1;
        int ans = 1;
        ans += rec(i-1,j);
        ans += rec(i,j-1);
        ans += rec(i+1,j);
        ans += rec(i,j+1);
        return ans;
    }
    
    int maxAreaOfIsland(vector<vector<int>>& grid) 
    {
        a = grid;
        m = a.size();
        n = a[0].size();
        
        vis.resize(m);
        for(auto &v:vis)
            v.resize(n);
        
        int ans = 0;
        for(int i=0;i<m;i++)
        {
            for(int j=0;j<n;j++)
            {
                if(a[i][j]==1 && !vis[i][j])
                {
                    ans = max(ans,rec(i,j));
                }
            }
        }
        return ans;
    }
};
